hal/src/rtl872x/backup_ram_hal.cpp

```diff
@@ -87,7 +91,15 @@ int hal_backup_ram_init(void) {
     // NOTE: using SDK API here as last reset info in core_hal is initialized later
     platform_system_flags_t dctFlags = {};
     dct_read_app_data_copy(DCT_SYSTEM_FLAGS_OFFSET, &dctFlags, DCT_SYSTEM_FLAGS_SIZE);
-    if ((BOOT_Reason() & BIT_BOOT_DSLP_RESET_HAPPEN) || SYSTEM_FLAG(restore_backup_ram) == 1 || dctFlags.restore_backup_ram == 1) {
+    // In case of software reset or watchdog reset (software reset also uses watchdog reset),
+    // RAM is preserved and we may end up restoring stale data out of flash instead.
+    // Check g_backupRamValidMarker value which resides in system retained area, if it's valid
+    // our backup RAM is fine as-is and we shouldn't restore anything.
+    SCOPE_GUARD({
+        g_backupRamValidMarker = HAL_BACKUP_RAM_VALID_VALUE;
```

Comment:
This block ensures that when `hal_backup_ram_init()` exits, `g_backupRamValidMarker` will always be set to `HAL_BACKUP_RAM_VALID_VALUE`. 
So `g_backupRamValidMarker` will only be reset if system RAM is lost (like when we sleep), which is the indication that we *DO* need to restore backup ram. We check for this below. 
Do I understand correctly?

Reply:
Yeah. Doing this on exit with `SCOPE_GUARD()` simply because of `CHECK()` potentially interrupting the flow of execution.